### DISCUSSION TITLE
Some fast prototyping Unreal+JSBSIM: F-15E (me) intercepting AI F-15E

### DISCUSSION THREAD

**Discussion by tbuckup** [JSBSim-Team/jsbsim#636] [Show and tell]
Hi guys, 

First post here, but a long time "follower". The recent plugin for Unreal triggered my interest. I spent the last 2 weeks learning a little bit of Unreal and put a tiny prototype with myself and AI flying over a region here in Germany (heightmap from SRTM ). You can see the video **[HERE](https://youtu.be/3tFF2iyK0DM)** in youtube. 
You will find more info/specs on description of the video. Perhaps interesting for other guys playing with that, JSBSIM component should run at fixed time step and not with the tick of game. I set 100Hz, which worked out ok as far as I could say when I increased the graphics and FPS dropped.

I am very interested in continuing the development of a serious combat flight simulator using Unreal and Jsbsim.  However, the jsbsim plugin (as already discussed on another thread here by @ICanDoThings2 and @AlbanBERGERET-Epic) is still complete. Among many things still missing, the lack of ability of sending back to the plugin information e.g. the weight of ordnance spent etc is a game breaker for a simulation that I have in mind.

If I understood correctly, @ICanDoThings2 is planning to continue development!?  I definitely would like to help advancing that, but I still have to learn how to create the plugins for unreal. 

Regards

Tiago aka tbuckup

**Comment by seanmcleod** [comment]
> the lack of ability of sending back to the plugin information e.g. the weight of ordnance spent etc is a game breaker for a simulation that I have in mind.

Why do you say there is no ability to get the weight of ordinance spent? The plug-in can set, e.g. to release ordinance and query the point mass properties in JSBSim via the reference it has to the JSBSim instance.

**Reply by tbuckup** [reply]
I missed this one, then. And you set ordnance via XML? Or I can do that in runtime via Unreal?

**Comment by seanmcleod** [comment]
JSBSim has no specific notion of ordinance. But supports the concept of point masses that can be placed on the aircraft. Look up references to point masses.

**Reply by tbuckup** [reply]
OK. It is via a definition on the XML file. It needs to be exposed then to the plugin.
Is it possible to define Cd for each point mass? In military aviation, each ordnance has a specific drag index (which is based on the nominal drag index of aircraft). I guess one could change the drag in run time when a drop mass is dropped.

**Comment by seanmcleod** [comment]
> It needs to be exposed then to the plugin.

The plug-in can access it via the exposed point mass properties in the property tree, see some mention of it here -https://github.com/JSBSim-Team/jsbsim/discussions/621.

> Is it possible to define Cd for each point mass?

No, the point mass doesn't have a $C_D$ field.

https://github.com/JSBSim-Team/jsbsim/blob/59389f66c790d0a8ae2323dbedbbf9b278e32e2c/src/models/FGMassBalance.h#L236-L241

So you would need to come up with a separate scheme to update the drag based on the stores, e.g. here is one scheme from @Zaretto's F-15 model. 

https://github.com/Zaretto/fg-aircraft/blob/d6ca80195169a8a0f333270f289f1f52db2b638b/aircraft/F-15/f15a.xml#L814-L839

**Reply by tbuckup** [reply]
The way Zaretto does is based on the XML. I do not want to use it, since I want to add stores inside unreal, pass to JSBSIM via the component in runtime and let it run. Drag for stores are known for acft and one can find them in the dash -34 manuals of USAF planes, for example. 
Here is an explanation on how the Drag Index (DI) is calculated: https://www.f-16.net/forum/viewtopic.php?t=14939
I will simply add the DI for each store using a look-up table in UE, then read the Cd table for the given acft and add the DI accordingly. 
The DIs of almost all stores for over 50 decades of planes are available in DCS and BMS Falcon 4. 
Let me learn how to write a plugin first for UE and then lets see if it works.....

**Reply by Zaretto** [reply]
The drag index for determining relative cruise performance; and as such there are a number of variables that aren't visible; such as the alpha - so we can only really use the drag index maybe as a multiplier on overall drag if we rescale appropriately. Drag index is also airframe specific.  Drag Index alone isn't sufficient to provide aerodynamics effects from the stores; for this you will need to have lookup tables for all coefficients (Lift, Drag, Side, Roll, Pitch, Yaw) and these will need to ideally be provided for (alpha, beta) - sideslip will have an effect.

In my F-15 I have two sets of tables one for tanks and the other for missiles. I work out the loading and this gives me a factor that is then applied to the coefficient. The factor is 0 for nothing and increments for the relative values for the stores.  This my compromise which gives an easily calculated value together with reasonable aerodynamic modelling; e.g. when asymmetric there will be a effect in yaw and roll.

**Reply by tbuckup** [reply]
It will be a simple DI...no tables.

**Comment by seanmcleod** [comment]
It also depends on the level of fidelity you're trying to achieve.

> Remembering that drag index is supposed to be used only to **assist in estimating cruise drag effects**, profile drag is the dominant type of drag in cruise condition.

Often for estimating cruise fuel burn etc.

Whereas if you take a look at @Zaretto's data, it's a lot more detailed for use by a 6DOF simulation with drag based on alpha and beta for the store. And if you look at the following section it even has the side-force generated by the store again indexed by alpha and beta. 

> then read the Cd table for the given acft and add the DI accordingly

Are you planning on trying to modify this sort of table at runtime after JSBSim has read it in?

https://github.com/JSBSim-Team/jsbsim/blob/59389f66c790d0a8ae2323dbedbbf9b278e32e2c/aircraft/f15/f15.xml#L488-L514

I'd suggest an easier approach in terms of adding a new element in the `Drag` section which adds additional drag based on the DI using a property you create, e.g.

```xml
<property>stores/DragIndex</property>
```

Then as the user adds stores and as stores are jettisoned you simply update your `stores/DragIndex` property.

You can decide what sort of fidelity you want to apply, e.g. very simple with just additional drag independent of alpha and beta, no side-force increments etc. or something as detailed as @Zaretto's model.

**Comment by ghost** [comment]
What I had planned and started working on, was not exactly a port, but rather an at least partial re-write and integration more closely with UE5. I really found the JSBSim plugin lacking. I know you want to make a Combat Flight Sim, and that's  what I wanted to do to an extent as well but there were other elements I wanted changed/added, etc too. For one, you're right if I understand you correctly, and there is no real way to remove/add weights at runtime, and that was one thing I was working on.

For an example in my project, I was wanting to implement drop tanks, which would obviously greatly increase the range of a given aircraft, however when depleted I believe per their namesake they're supposed to detach and fall away from the airplane, which would reduce the mass fairly noticeably and it seems JSBSim, or the plugin of it for UE5 didn't really support this. There would also be other reasons this would be of interest to me personally at least, for say a cargo plane which would change mass, or even just one that dropped off paratroopers.

Furthermore, I also found that the UE5 plugin still had no support for other aircrafts than jets, which is a big issue if you want to do more than just the standard-fare for most, like even if you just want to sim a Cessna or something of that nature, and there are also no helicopters either. 

So in conclusion I had a fair amount of reasons for doing what I was trying to, however, I put it on pause for other reasons not that I have discontinued it.

**Comment by seanmcleod** [comment]
> there is no real way to remove/add weights at runtime

You can use point-masses, and create one for each station under the wings, wing tips and fuselage. You can adjust the weight of each one at run time, see https://github.com/JSBSim-Team/jsbsim/discussions/621 either to load/add an ordinance, drop tank etc. or to drop/jettison one.

**Reply by ghost** [reply]
Technically yes, however there is the question about flexibility at runtime which is kind of reduced by having a set number of points instead of being able to add/remove them at runtime.

**Reply by bcoconni** [reply]
Well, if you want to dynamically add/remove point masses via C++ calls, it should be quite easy to add an alternative method `FGMassBalance::AddPointMass` that is inspired by the currently existing one:
https://github.com/JSBSim-Team/jsbsim/blob/59389f66c790d0a8ae2323dbedbbf9b278e32e2c/src/models/FGMassBalance.cpp#L266-L320
The current method `FGMassBalance::AddPointMass` is expecting some XML as an input, but a new method could use a set of C++ parameters that would supply the data that is currently supplied by XML.

And its sibling `FGMassBalance::RemovePointMass` would be straightforward to code.

If these methods would help meeting your goal, feel free to submit a pull request.

**Comment by seanmcleod** [comment]
> by having a set number of points

But each aircraft, e.g. an F-15, F-16 has a fixed number of stations. So why not set the specific stations for the aircraft type in it's xml file with the XYZ location for each one. Then at runtime based on user input in terms of the loadout they want to start off with you then simply update the weight from 0 to whatever weight is relevant for the particular item, and set it's weight to 0 when the user drops/jettisons/launches an item.

**Reply by ghost** [reply]
Fair, but my point is something like a cargo plane won't. It will have space inside the fuselage you can tie the cargo down at any point in. In any case, there were other changes I felt worthy of being made and I was making it more integrated/coupled with UE5.

**Reply by tbuckup** [reply]
@seanmcleod Basically what I intend to do. 
my to-do list is to add a "sockets" to each airframe. This socket can be a pylon, but also cargo, even a space shuttle over  a 747. They will have a position (hard coded on the xml-file), weight and DI (simple value, no tables), which will be updated in real time from UE5. I will have to do some changes on jsbsim. I may need your help here.
I still did not have time to learn how to write a plugin to communicate with jsbsim...:-(

**Comment by gallonmate** [comment]
well until the license issues are solved, as brought up in another discussion, i wouldn't put all your heart into this plugin yet. I don't mean stop your own development. It depends on your final plans and end usage of your game/sim, but all it takes is one of the many contributors to JSBSim project to complain and your project is potentially dead. I'm sure the issues will be solved soon but there's a chance the plugin will also be different by the time it meets all of JSBSim license requirements and even requirements for Flight model/xml with GNU licenses.

**Reply by ghost** [reply]
What license issues are you referring to?

**Reply by gallonmate** [reply]
https://github.com/JSBSim-Team/jsbsim/discussions/631
